savant/deepstream/source_output_factory.py

```diff
@@ -0,0 +1,214 @@
+                [
+                    'video/x-raw(memory:NVMM)',
+                    'format=RGBA',
+                    f'width={source_info.dest_resolution.width}',
+                    f'height={source_info.dest_resolution.height}',
+                ]
+            )
+        )
+
+
+class SourceOutputEncoded(SourceOutputFactoryWithFrame):
+    """Factory for adding output elements to a DeepStream pipeline.
+    Output contains encoded frames along with metadata.
+    """
+
+    def __init__(
+        self,
+        codec: CodecInfo,
+        params: Dict[str, Any] = None,
+    ):
+        super().__init__()
+        self._codec = codec
+        self._params = params or {}
+
+    def _add_transform_elems(self, pipeline: GstPipeline, source_info: SourceInfo):
+        encoder = pipeline._add_element(
+            PipelineElement(self._codec.encoder, properties=self._params)
+        )
+        source_info.after_demuxer.append(encoder)
+        encoder.sync_state_with_parent()
+        self._logger.debug(
+            'Added encoder %s with parms %s', self._codec.encoder, self._params
+        )
+
+    def _build_output_caps(self, source_info: SourceInfo) -> Gst.Caps:
+        return Gst.Caps.from_string(
+            ', '.join(
+                [
+                    self._codec.caps_name,
+                    f'width={source_info.dest_resolution.width}',
+                    f'height={source_info.dest_resolution.height}',
+                ]
+            )
+        )
+
+
+class SourceOutputH26X(SourceOutputEncoded):
+    """Factory for adding output elements to a DeepStream pipeline.
+    Output contains frames encoded with h264 or h265 (hevc) codecs along with metadata.
+    """
+
+    def _add_transform_elems(self, pipeline: GstPipeline, source_info: SourceInfo):
+        super()._add_transform_elems(pipeline, source_info)
+        # A parser for codecs h264, h265 is added to include
+        # the Sequence Parameter Set (SPS) and the Picture Parameter Set (PPS)
+        # to IDR frames in the video stream. SPS and PPS are needed
+        # to correct recording or playback not from the beginning
+        # of the video stream.
+        parser_params = {'config-interval': -1}
+        parser = pipeline._add_element(
+            PipelineElement(self._codec.parser, properties=parser_params)
+        )
+        source_info.after_demuxer.append(parser)
+        parser.sync_state_with_parent()
+        self._logger.debug(
+            'Added parser %s with parms %s', self._codec.parser, parser_params
+        )
+
+
+class SourceOutputPng(SourceOutputEncoded):
+    """Factory for adding output elements to a DeepStream pipeline.
+    Output contains frames encoded with PNG codec along with metadata.
+    """
+
+    def dest_resolution(self, source_info: SourceInfo) -> Resolution:
+        # Rounding resolution to the multiple of 8 to avoid cuda errors.
+        return Resolution(
+            width=round(source_info.src_resolution.width / 8) * 8,
```

Comment:
Resolution that is not multiple of 8 causes cuda errors.

```
nvbufsurface: NvBufSurfaceSysToHWCopy: failed in mem copy
nvbufsurface: NvBufSurfaceCopy: failed to copy
Unable to release device memory.
Unable to release host memory.
Cuda failure: status=700
nvbufsurface: Error(-1) in releasing cuda memory
```

It can be reproduced by using `pictures` and passing a directory with two images to it: the first image has resolution `1848x1173` and the second one `1266x1068`.

@denisvmedyantsev Not sure if we need to round resolution for any codec or only for PNG. I didn't run into this issue on other codecs.

Reply:
Probably it's that bug: https://forums.developer.nvidia.com/t/nvstreammux-error-releasing-cuda-memory/219895

I'll try to use `interpolation-method=6` on `nvvideoconvert`.

Reply:
`interpolation-method=6` on `nvvideoconvert` doesn't help.